src/qibo/tensorflow/gates.py

```diff
@@ -35,6 +36,22 @@ def _prepare(self):
         rank = int(np.log2(int(matrix.shape[0])))
         self.matrix = tf.reshape(matrix, 2 * rank * (2,))
 
+    def _calculate_einsum_cache(self):
+        if self.is_controlled_by:
+            if self.density_matrix:
+                from qibo.tensorflow import einsum
+                self.einsum = einsum.DefaultEinsum()
```

Comment:
Could you please remember me why? (maybe we need some comments here)

Reply:
After expanding the density matrix tests in the two previous PRs, I noticed that `'matmuleinsum'` was not working properly for the case where `controlled_by` gates were used on density matrices. This is a temporary solution that falls back to the `'defaulteinsum'` automatically so that everything works properly. The long term solution would be to actually fix matmuleinsum. I just did not want to delay these PRs for this because I believe the combination density matrices + controlled by + matmuleinsum would be very rare (if it ever appears) in practice. Perhaps I should open an issue about that though.